File: jvm-experiments-timeperiod/src/main/java/kr/kth/timeperiod/timerange/DayRange.java
package kr.kth.timeperiod.timerange;

import kr.kth.commons.Guard;
import kr.kth.timeperiod.ITimeCalendar;
import kr.kth.timeperiod.ITimeFormatter;
import kr.kth.timeperiod.TimeCalendar;
import kr.kth.timeperiod.TimeFormatter;
import kr.kth.timeperiod.clock.ClockProxy;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;

/**
 * 1일 기간을 나타내는 클래스입니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 12. 25.
 */
@Slf4j
public class DayRange extends DayTimeRange {

    private static final long serialVersionUID = 5495008766913620838L;

    // region << Constructors >>

    public DayRange() {
        this(TimeCalendar.Default());
    }

    public DayRange(ITimeCalendar calendar) {
        this(ClockProxy.getClock().now(), calendar);
    }

    public DayRange(DateTime moment) {
        this(moment, TimeCalendar.Default());
    }

    public DayRange(DateTime moment, ITimeCalendar calendar) {
        super(moment, 1, calendar);
    }

    public DayRange(int year, int month, int day) {
        this(year, month, day, TimeCalendar.Default());
    }

    public DayRange(int year, int month, int day, ITimeCalendar calendar) {
        super(year, month, day, 1, calendar);
    }

    // endregion

    public int getYear() {
        return getStartYear();
    }

    public int getMonth() {
        return getStartMonth();
    }

    public int getDay() {
        return getStartDay();
    }

    public int getDayOfWeek() {
        return getStartDayOfWeek();
    }

    public String getDayName() {
        return getStartDayName();
    }

    public HourRange getPreviousDay() {
        return addDays(-1);
    }

    public HourRange getNextDay() {
        return addDays(1);
    }

    public HourRange addDays(int days) {
        DateTime start = getStart().withTimeAtStartOfDay();
        return new HourRange(start.plusDays(days), getTimeCalendar());
    }

    @Override
    protected String format(ITimeFormatter formatter) {
        ITimeFormatter fmt = Guard.firstNotNull(formatter, TimeFormatter.getInstance());

        return fmt.getCalendarPeriod(fmt.getShortDate(getStart()),
                fmt.getShortTime(getStart()),
                fmt.getShortTime(getEnd()),
                getDuration());
    }
}

File: jvm-experiments-timeperiod/src/main/scala/kr/kth/timeperiod/timerange/DayRange.scala
package kr.kth.timeperiod.timerange

import org.joda.time.DateTime
import kr.kth.timeperiod._
import DayOfWeek._
import kr.kth.commons.tools.ScalaHash
import scala.collection.immutable.IndexedSeq

/**
 * 일단위 기간을 표현하는 추상 클래스입니다.
 * @param start
 * @param dayCount
 * @param calendar
 */
abstract class DayTimeRange(start: DateTime, dayCount: Int, calendar: ITimeCalendar)
	extends CalendarTimeRange(start, start.plusDays(dayCount), calendar) {

	private val _dayCount = dayCount

	def this(year: Int, monthOfYear: Int, dayOfMonth: Int, dayCount: Int, calendar: ITimeCalendar) {
		this(new DateTime().withDate(year, monthOfYear, dayOfMonth), dayCount, calendar)
	}

	def this(year: Int, monthOfYear: Int, dayOfMonth: Int, dayCount: Int) {
		this(year, monthOfYear, dayOfMonth, dayCount, TimeCalendar.Default)
	}

	def getDayCount = _dayCount

	def getStartDayOfWeek: DayOfWeek = getTimeCalendar.getDayOfWeek(getStart)

	def getStartDayName: String = getTimeCalendar.getDayName(getStartDayOfWeek)

	def getEndDayOfWeek: DayOfWeek = getTimeCalendar.getDayOfWeek(getEnd)

	def getEndDayName: String = getTimeCalendar.getDayName(getEndDayOfWeek)

	def getHours: Iterable[HourRange] = {
		val startDay = getStartDayStart
		val calendar = getTimeCalendar

		for {
			d <- (0 until getDayCount)
			h <- (0 until TimeSpec.HoursPerDay)
		}
		yield new HourRange(startDay.plusDays(d).plusHours(h), calendar)
	}

	override def hashCode: Int = ScalaHash.compute(super.hashCode, getDayCount)

	protected override def buildStringHelper() =
		super.buildStringHelper()
		.add("dayCount", getDayCount)
}

/**
 * 일단위 기간을 표현합니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 13. 1. 17.
 */
class DayRange(moment: DateTime, calendar: ITimeCalendar) extends DayTimeRange(moment, 1, calendar) {

	def this(calendar: ITimeCalendar) {
		this(Clock.getClock.now, calendar)
	}

	def this(moment: DateTime) {
		this(moment, TimeCalendar.Default)
	}

	def this() {
		this(TimeCalendar.Default)
	}

	def this(year: Int, monthOfYear: Int, dayOfMonth: Int, calendar: ITimeCalendar) {
		this(new DateTime(year, monthOfYear, dayOfMonth, 0, 0), calendar)
	}

	def this(year: Int, monthOfYear: Int, dayOfMonth: Int) {
		this(new DateTime(year, monthOfYear, dayOfMonth, 0, 0))
	}

	def getYear = getStartYear

	def getMonth = getStartMonth

	def getDay = getStartDay

	def getDayOfWeek: DayOfWeek = getStartDayOfWeek

	def getDayName = getStartDayName

	def getPreviousDay = addDays(-1)

	def getNextDay = addDays(1)

	def addDays(days: Int): DayRange = {
		new DayRange(Times.getDatePart(getStart).plusDays(days), getTimeCalendar)
	}

	override def format(formatter: Option[ITimeFormatter]) = {
		val fmt = formatter.getOrElse(TimeFormatter.instance)
		fmt.getCalendarPeriod(getDayName,
		                      fmt.getShortDate(getStart),
		                      fmt.getShortDate(getEnd),
		                      getDuration)
	}
}

/**
 * 일단위 기간을 표현하는 `DayRange` 의 컬렉션을 표현합니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 13. 1. 17.
 */
class DayRangeCollection(moment: DateTime, dayCount: Int, calendar: ITimeCalendar)
	extends DayTimeRange(moment, dayCount, calendar) {

	def this(moment: DateTime, dayCount: Int) {
		this(moment, dayCount, TimeCalendar.Default)
	}

	def this(year: Int, monthOfYear: Int, dayOfMonth: Int, dayCount: Int, calendar: ITimeCalendar) {
		this(new DateTime().withDate(year, monthOfYear, dayOfMonth), dayCount, calendar)
	}

	def this(year: Int, monthOfYear: Int, dayOfMonth: Int, dayCount: Int) {
		this(year, monthOfYear, dayOfMonth, dayCount, TimeCalendar.Default)
	}

	def getDays: IndexedSeq[DayRange] = {
		val startDay = getStart

		(0 until getDayCount).map(d => new DayRange(startDay.plusDays(d), getTimeCalendar))
	}

	override def format(formatter: Option[ITimeFormatter]) = {
		val fmt = formatter.getOrElse(TimeFormatter.instance)
		fmt.getCalendarPeriod(getStartDayName,
		                      getEndDayName,
		                      fmt.getShortDate(getStart),
		                      fmt.getShortDate(getEnd),
		                      getDuration)

	}
}